address/address.go

```diff
@@ -47,327 +77,236 @@ const (
-func ParseError(addr string) error {
-	hrp, version, data, err := decode(addr)
+// UnmarshalJSON implements the json unmarshal interface.
+func (a *Address) UnmarshalJSON(b []byte) error {
+	in := strings.TrimSuffix(strings.TrimPrefix(string(b), `"`), `"`)
+	addr, err := decode(in)
 	if err != nil {
-		return errors.Wrap(err, "unable to decode")
-	}
-
-	if _, err := NetworkFromString(hrp); err != nil {
-		return errors.Wrap(err, "invalid network")
-	}
-
-	if version != Version {
-		return fmt.Errorf("invalid version: version=%d", version)
-	}
-
-	if len(data) != HashLength {
-		return fmt.Errorf("invalid data length: len=%d", len(data))
+		return err
 	}
-
+	*a = addr
 	return nil
 }
 
-// Empty returns true if the address is empty.
-func (a Address) Empty() bool {
-	return a == (Address{})
+// MarshalJSON implements the json marshal interface.
+func (a Address) MarshalJSON() ([]byte, error) {
+	return json.Marshal(a.String())
 }
 
-// Network returns the network of the address.
-func (a Address) Network() Network {
-	return a[0]
+// NewIDAddress returns an address using the ID protocol.
+func NewIDAddress(id uint64) Address {
+	return newAddress(ID, leb128.FromUInt64(id))
 }
 
-// Version returns the version of the address.
-func (a Address) Version() byte {
-	return a[1]
+// NewSecp256k1Address returns an address using the SECP256K1 protocol.
+func NewSecp256k1Address(pubkey []byte) Address {
+	return newAddress(SECP256K1, addressHash(pubkey))
 }
 
-// Hash returns the hash part of the address.
-func (a Address) Hash() []byte {
-	return a[2:]
+// NewActorAddress returns an address using the Actor protocol.
+func NewActorAddress(data []byte) Address {
+	return newAddress(Actor, addressHash(data))
 }
 
-// Format implements the Formatter interface.
-func (a Address) Format(f fmt.State, c rune) {
-	switch c {
-	case 'v':
-		fmt.Fprintf(f, "[%s - %x - %x]", NetworkToString(a.Network()), a.Version(), a.Hash()) // nolint: errcheck
-	case 's':
-		fmt.Fprintf(f, "%s", a.String()) // nolint: errcheck
-	default:
-		fmt.Fprintf(f, "%"+string(c), a.Bytes()) // nolint: errcheck
-	}
+// NewBLSAddress returns an address using the BLS protocol.
+func NewBLSAddress(pubkey []byte) Address {
+	return newAddress(BLS, pubkey)
 }
 
-// MarshalText implements the TextMarshaler interface.
-func (a Address) MarshalText() ([]byte, error) {
-	if a == (Address{}) {
-		return nil, nil
-	}
-
-	return []byte(a.String()), nil
+// NewFromString returns the address represented by the string `addr`.
+func NewFromString(addr string) (Address, error) {
+	return decode(addr)
 }
 
-// UnmarshalText implements the TextUnmarshaler interface.
-func (a *Address) UnmarshalText(in []byte) error {
-	if len(in) == 0 {
-		return nil
-	}
-
-	out, err := NewFromString(string(in))
-	if err != nil {
-		return err
-	}
-
-	copy(a[:], out[:])
-
-	return nil
+// NewFromBytes return the address represented by the bytes `addr`.
+func NewFromBytes(addr []byte) Address {
+	return Address{string(addr)}
```

Comment:
definitely need some validation here, esp if this is coming from potentially untrusted sources.

Reply:
agreed, will address.